File: OOPS/Inheritance/access_specifiers.py
'''
Python allows the accessibility of all members from anywhere publicly.
There is no private or protected access specifier.

But, 

i) _ (single underscore) is used to denote the private members. It does not affect the accessibility. 
    [A convention that the members should not be used from outside the class]

ii) __ (double underscore) restricts direct access to the member from the instance.
    But the member can be accessed using the class name.

'''

#Base class
class Parent:

    def _printName(self):
        print("Single Underscore method")

    def __printName(self):
        print("Double Underscore method")

#Parent class
class Child(Parent):

    #inherits all members from Parent
    pass


c = Child()
c._printName()
c._Parent__printName()


'''
Output:

Single Underscore method
Double Underscore method
'''

File: OOPS/Inheritance/access_specifiers.cpp
/*
Access specifiers or Access modifiers:

    - Define the accessibility of the class members
    - There are three types of access specifiers
        i) public
       ii) private
      iii) protected
    - Default access specifier is private
        

public: 

    If the child class inherits from parent class using public mode, then
    
    | Base Class           |      Child Class         |
    | ----------------------------------------------- |
    | public members    ---|--->  public members      |
    | protected members ---|--->  protected members   |
    | private members   ---|--->  Can't be inherited  |
    

protected: 

    If the child class inherits from parent class using protected mode, then
    
    | Base Class           |      Child Class         |
    | ----------------------------------------------- |
    | public members    ---|--->  protected members   |
    | protected members ---|--->  protected members   |
    | private members   ---|--->  Can't be inherited  |


private:

    If the child class inhertis from parent class using private mode, then

    | Base Class           |      Child Class         |
    | ----------------------------------------------- |
    | public members    ---|--->  private members     |
    | protected members ---|--->  private members     |
    | private members   ---|--->  Can't be inherited  |

*/


#include<iostream>
using namespace std;

//Parent class
class Parent{

    public:
        int A=10;
    protected:
        int B=20;
    private:
        int C=30;

};  

//inherits using public specifier
class ChildPublic : public Parent{

    /*
        A is inherited as public
        B is inherited as protected
        C can not be inherited
    */

};

//inherits using protected specifier
class ChildProtected: protected Parent{

    /*
        A is inherited as protected
        B is inherited as protected
        C can not be inherited
    */

};

//inherits using private specifier
class ChildPrivate: private Parent{

    /*
        A is inherited as private
        B is inherited as private
        C can not be inherited
    */

};

int main(){

    return 0;
}